File: 1969.minimum-non-zero-product-of-the-array-elements.cpp
/*
 * @lc app=leetcode id=1969 lang=cpp
 *
 * [1969] Minimum Non-Zero Product of the Array Elements
 */

// @lc code=start
class Solution {
public:
    int modPow(long long x, long long y, int m)
    {
        if (y == 0)
            return 1;
        long long p = modPow(x, y / 2, m);
        p = (p * p) % m;
        return y % 2 ? (p * (x % m)) % m : p;
    }
    int minNonZeroProduct(int p) {
        long long cnt = (1ll << p) - 1, mod = 1000000007;
        return cnt % mod * modPow(cnt - 1, cnt / 2, mod) % mod;
    }
};
// @lc code=end

File: 1969.minimum-non-zero-product-of-the-array-elements.cs
/*
 * @lc app=leetcode id=1969 lang=csharp
 *
 * [1969] Minimum Non-Zero Product of the Array Elements
 */

// @lc code=start
public class Solution {
    public int ModPow(long x, long y, int m) {
        if (y == 0)
            return 1;
        long p = ModPow(x, y / 2, m);
        p = (p * p) % m;
        return y % 2 == 1 ? (int)((p * (x % m)) % m) : (int)p;
    }

    public int MinNonZeroProduct(int p) {
        long cnt = (1L << p) - 1;
        int mod = 1000000007;
        return (int)((cnt % mod) * ModPow(cnt - 1, cnt / 2, mod) % mod);
    }
}


// @lc code=end

